src/Storages/StorageInMemoryMetadata.cpp

```diff
@@ -81,6 +90,63 @@ void StorageInMemoryMetadata::setComment(const String & comment_)
     comment = comment_;
 }
 
+void StorageInMemoryMetadata::setDefiner(const ASTSQLSecurity & sql_security)
+{
+    if (sql_security.definer)
+        definer = sql_security.definer->toString();
+
+    sql_security_type = sql_security.type;
+}
+
+UUID StorageInMemoryMetadata::getDefinerID(DB::ContextPtr context) const
+{
+    if (!definer)
+    {
+        if (const auto definer_id = context->getUserID())
+            return *definer_id;
+
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "No user in context for sub query execution.");
+    }
+
+    const auto & access_control = context->getAccessControl();
+    return access_control.getID<User>(*definer);
+}
+
+ContextMutablePtr StorageInMemoryMetadata::getSQLSecurityOverriddenContext(ContextPtr context) const
+{
+    if (!sql_security_type.has_value())
+        return Context::createCopy(context);
+
+    if (sql_security_type == ASTSQLSecurity::Type::INVOKER)
+        return Context::createCopy(context);
+
+    auto new_context = Context::createCopy(context->getGlobalContext());
+    new_context->applySettingsChanges(context->getSettingsRef().changes());
+    new_context->setClientInfo(context->getClientInfo());
+    if (context->hasQueryContext())
+        new_context->setQueryContext(context->getQueryContext());
+
+    const auto & database = context->getCurrentDatabase();
```

Comment:
How can `context->getCurrentDatabase()` be empty?

Reply:
I don't know, but it can (found via tests) :sweat_smile: